autoprecompiles/src/optimizer.rs

```diff
@@ -112,60 +105,79 @@ fn solver_based_optimization<T: FieldElement>(
+        .cloned()
+        .collect::<HashSet<_>>();
+
+    // Any variable that is connected to a variable in vars_to_keep must also be kept.
+    loop {
+        let size_before = variables_to_keep.len();
+        for constraint in &constraint_system.algebraic_constraints {
+            if constraint
+                .referenced_variables()
+                .any(|var| variables_to_keep.contains(var))
+            {
+                variables_to_keep.extend(constraint.referenced_variables().cloned());
+            }
+        }
+
+        for constraint in &constraint_system.bus_interactions {
+            if constraint
+                .referenced_variables()
+                .any(|var| variables_to_keep.contains(var))
+            {
+                variables_to_keep.extend(constraint.referenced_variables().cloned());
+            }
+        }
+        if variables_to_keep.len() == size_before {
+            break;
+        }
+    }
 
     ConstraintSystem {
-        algebraic_constraints,
-        bus_interactions: bus_interactions
+        algebraic_constraints: constraint_system
+            .algebraic_constraints
+            .into_iter()
+            .filter(|constraint| {
+                constraint
+                    .referenced_variables()
+                    .any(|var| variables_to_keep.contains(var))
+            })
+            .collect(),
+        bus_interactions: constraint_system
+            .bus_interactions
             .into_iter()
-            .filter_map(|bus_interaction| {
-                if let Some(concrete_bus_interaction) =
-                    try_to_concrete_bus_interaction(&bus_interaction)
-                {
-                    // If all values are concrete, we might be able to remove the bus interaction
-                    match bus_interaction_handler
-                        .handle_concrete_bus_interaction(concrete_bus_interaction)
-                    {
-                        ConcreteBusInteractionResult::AlwaysSatisfied => None,
-                        ConcreteBusInteractionResult::HasSideEffects => Some(bus_interaction), // Here we still keep the original bus interaction
-                        ConcreteBusInteractionResult::ViolatesBusRules => {
-                            panic!("Bus interaction {bus_interaction:?} violates bus rules");
-                        }
-                    }
-                } else {
-                    // If any value is symbolic, we keep the bus interaction
-                    Some(bus_interaction)
-                }
+            .filter(|bus_interaction| {
+                let bus_id = bus_interaction.bus_id.try_to_number().unwrap();
+                let has_vars_to_keep = bus_interaction
+                    .referenced_variables()
+                    .any(|var| variables_to_keep.contains(var));
+                // has_vars_to_keep would also be true for bus interactions containing only
```

Comment:
huh, why? Do you mean the opposite, that is, that `has_vars_to_keep` would be false for the only constants case, which would remove the bus interaction, but you want to keep the stateful ones?